--- a/drivers/of/of_net.c
+++ b/drivers/of/of_net.c
@@ -11,6 +11,9 @@
 #include <linux/phy.h>
 #include <linux/export.h>
 #include <linux/mtd/mtd.h>
+
+#define WT_INFO_MTD_NAME		"wtinfo"
+extern void wt_get_mac(void *mac);
 
 /**
  * of_get_phy_mode - Get phy mode for given device_node
@@ -78,12 +81,16 @@
 	if (!part)
 		part = mtd_np->name;
 
-	mtd = get_mtd_device_nm(part);
-	if (IS_ERR(mtd))
-		return NULL;
+	if (strncmp(part, WT_INFO_MTD_NAME, strlen(WT_INFO_MTD_NAME))) {
+		mtd = get_mtd_device_nm(part);
+		if (IS_ERR(mtd))
+			return NULL;
 
-	ret = mtd_read(mtd, be32_to_cpup(list), 6, &retlen, mac);
-	put_mtd_device(mtd);
+		ret = mtd_read(mtd, be32_to_cpup(list), 6, &retlen, mac);
+		put_mtd_device(mtd);
+	} else {
+		wt_get_mac((void *)mac);
+	}
 
 	if (!of_property_read_u32(np, "mtd-mac-address-increment", &mac_inc))
 		mac[5] += mac_inc;
